### ISSUE TITLE
Issue with parsing expression with unary operator for macro rules

### ISSUE THREAD

**Issue by AckslD** [rust-lang/rust#82968]
It seems to be an issue with `literal`s vs `expr`s in macro rules, in particular for the unary operation `-`.

I tried this code:

```rust
#[macro_export]
macro_rules! test {
    ($a:literal) => { $a };
    ($a:expr) => { $a };
}

fn func() {
    let a = test!(0);
    let b = test!(a);
    let c = test!(-a);
}
```

I expected to see this happen: No errors and that the second branch would be used for the assignment of both `b` and `c`.

Instead, this happened: I got the following error:
```
error: unexpected token: `a`
   --> src/gate.rs:144:20
    |
137 |     ($a:literal) => { $a };
    |      ---------- while parsing argument for this `literal` macro fragment
...
144 |     let c = test!(-a);
    |     
```

### Meta
`rustc --version --verbose`:
```
rustc 1.52.0-nightly (3a5d45f68 2021-03-09)
binary: rustc
commit-hash: 3a5d45f68cadc8fff4fbb557780f92b403b19c19
commit-date: 2021-03-09
host: x86_64-unknown-linux-gnu
release: 1.52.0-nightly
LLVM version: 12.0.0
```

<details><summary>Backtrace</summary>
<p>

```
   Compiling rs-bug v0.1.0 (/home/axel/tmp/rs-bug)
error: unexpected token: `a`
  --> src/main.rs:10:20
   |
3  |     ($a:literal) => { $a };
   |      ---------- while parsing argument for this `literal` macro fragment
...
10 |     let c = test!(-a);
   |                    ^

error: aborting due to previous error

error: could not compile `rs-bug`

To learn more, run the command again with --verbose. 
```

</p>
</details>

**Comment by jonas-schievink** [comment]
This is working as intended, non-terminal metavariables always fully commit to parsing a part of the syntax, if this fails, the whole macro call is rejected.

**Comment by AckslD** [comment]
Thanks @jonas-schievink! I see now.

Although, there still seems to be an inconsistency. Since the above code gives an error, however, if I add parenthesis like `let c = test!((-a));` it works but gives the warning
```
warning: unnecessary parentheses around assigned value
   --> src/gate.rs:144:19
    |
144 |     let c = test!((-a));
    |                   ^^^^ help: remove these parentheses
    |
    = note: `#[warn(unused_parens)]` on by default
```
But clearly the parenthesis is not unnecessary.
